src/prof_recent.c

```diff
@@ -469,66 +580,65 @@ prof_recent_alloc_dump(tsd_t *tsd, void (*write_cb)(void *, const char *),
 	emitter_init(&emitter, emitter_output_json_compact,
 	    buf_writer_get_write_cb(&buf_writer),
 	    buf_writer_get_cbopaque(&buf_writer));
+
+	malloc_mutex_lock(tsd_tsdn(tsd), &prof_dump_mtx);
 	emitter_begin(&emitter);
 
 	malloc_mutex_lock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
-	prof_recent_alloc_assert_count(tsd);
+	ssize_t dump_count = prof_recent_alloc_dump_prepare_locked(tsd);
+	ssize_t dump_max = prof_recent_alloc_max_get(tsd);
+	malloc_mutex_unlock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
 
-	/*
-	 * Set prof_recent_alloc_max to 0 so that dumping won't block sampled
-	 * allocations: the allocations can complete but will not be recorded.
-	 */
-	ssize_t max = prof_recent_alloc_max_update(tsd, 0);
+	emitter_json_kv(&emitter, "recent_alloc_max", emitter_type_ssize,
+		&dump_max);
+	emitter_json_array_kv_begin(&emitter, "recent_alloc");
 
-	emitter_json_kv(&emitter, "recent_alloc_max", emitter_type_ssize, &max);
+	prof_recent_t *dump_last = NULL;
 
-	emitter_json_array_kv_begin(&emitter, "recent_alloc");
-	for (prof_recent_t *n = prof_recent_alloc_begin(tsd);
-	    n != prof_recent_alloc_end(tsd);
-	    n = prof_recent_alloc_next(tsd, n)) {
-		emitter_json_object_begin(&emitter);
-
-		emitter_json_kv(&emitter, "size", emitter_type_size, &n->size);
-		size_t usize = sz_s2u(n->size);
-		emitter_json_kv(&emitter, "usize", emitter_type_size, &usize);
-		bool released = n->alloc_edata == NULL;
-		emitter_json_kv(&emitter, "released", emitter_type_bool,
-		    &released);
-
-		emitter_json_kv(&emitter, "alloc_thread_uid",
-		    emitter_type_uint64, &n->alloc_tctx->thr_uid);
-		uint64_t alloc_time_ns = nstime_ns(&n->alloc_time);
-		emitter_json_kv(&emitter, "alloc_time", emitter_type_uint64,
-		    &alloc_time_ns);
-		emitter_json_array_kv_begin(&emitter, "alloc_trace");
-		dump_bt(&emitter, n->alloc_tctx);
-		emitter_json_array_end(&emitter);
-
-		if (n->dalloc_tctx != NULL) {
-			assert(released);
-			emitter_json_kv(&emitter, "dalloc_thread_uid",
-			    emitter_type_uint64, &n->dalloc_tctx->thr_uid);
-			assert(!nstime_equals_zero(&n->dalloc_time));
-			uint64_t dalloc_time_ns = nstime_ns(&n->dalloc_time);
-			emitter_json_kv(&emitter, "dalloc_time",
-			    emitter_type_uint64, &dalloc_time_ns);
-			emitter_json_array_kv_begin(&emitter, "dalloc_trace");
-			dump_bt(&emitter, n->dalloc_tctx);
-			emitter_json_array_end(&emitter);
-		} else {
-			assert(nstime_equals_zero(&n->dalloc_time));
+	buf_writer_set_allow_auto_flush(&buf_writer, false);
+	malloc_mutex_lock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+	prof_recent_t *dump_dummy = prof_recent_alloc_dummy_reserve;
+	for (prof_recent_t *node = dump_dummy->next; node != dump_dummy;
+	    node = node->next) {
+		buf_writer_commit(&buf_writer);
+		prof_recent_alloc_dump_node(&emitter, node);
+		if (buf_writer_overflow(&buf_writer)) {
```

Comment:
If overflowed, the current node is not written, and we should continue w/o the node=node->next?

Reply:
That's great catch! I was particularly paying attention to this case in the more complicated version, but forgot about it in this new version.